app/main.py

```diff
@@ -0,0 +1,97 @@
+import bcrypt,jwt
+from fastapi import FastAPI, status
+from fastapi.exceptions import HTTPException
+from sqlmodel import Session, select
+from starlette.middleware.cors import CORSMiddleware
+
+from db import engine, create_db_and_tables, SECRET_KEY, ALGORITHM
+from models.models import User, UserUpdate, Token, LoginRequest
+
+app = FastAPI()
+
+app.add_middleware(
+    CORSMiddleware,
+    allow_origins=["*"],
+    allow_credentials=True,
+    allow_methods=["*"],
+    allow_headers=["*"],
+)
+
+@app.post("/users/signup", status_code=status.HTTP_201_CREATED)
+def create_user(user:User):
+    with Session(engine) as session:
+        if session.exec(select(User).filter(User.email == user.email)).first():
+            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail = "user is exist")
+
+        user.password = bcrypt.hashpw(user.password.encode('utf-8'), bcrypt.gensalt()).decode() 
+        session.add(user)
+        session.commit()
+        session.close()
+        return {"message" : "create user"}
+ 
+@app.post("/users/login", response_model=Token)
+async def login_for_access_token(login_request: LoginRequest):
+    user = authenticate_user(email=login_request.email, password=login_request.password)
+    if not user:
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Incorrect email or password",
+            headers={"WWW-Authenticate": "Bearer"},
+        )
+   
+    access_token = create_access_token(email=user.email)
+
+    return {"access_token": access_token, "token_type": "bearer"}
+
+def authenticate_user(email: str, password: str):
+    with Session(engine) as session:
+        user = session.exec(select(User).filter(User.email == email)).first()
+
+    if not user:
+        return None
+    if not check_password(password, user.password):
+        return None
+    return user
+
+def create_access_token(email: str):
+    payload = {"email": email}
+    encoded_jwt = jwt.encode(payload, SECRET_KEY, algorithm=ALGORITHM)
+    return encoded_jwt
+
+def check_password(password:str, db_password:str):
+    checkpw = bcrypt.checkpw(password.encode('utf-8'), db_password.encode('utf-8'))
+    return checkpw 
+
+@app.patch("/users/{user_id}")
+def update_user(user_id : int, user:UserUpdate):
+    with Session(engine) as session:
+        db_user = session.exec(select(User).where(User.id == user_id)).one()
+        
+        if not db_user:
+            raise HTTPException(status_code = status.HTTP_404_NOT_FOUND, detail = "user not exist")
+
+        if user.email:
+            db_user.email = user.email
+
+        if user.password:
+            db_user.password = user.password
```

Comment:
P2: pydantic 모델에서 이미 입력필드와 값에 대한 검증을 하고 있어서 조건문은 삭제해도 될 것 같아요.
```suggestion
          db_user.email = user.email
          db_user.password = user.password
```

Reply:
네! 수정하였습니다!